File: domain/src/main/java/com/arturogutierrez/openticator/interactor/Interactor.java
package com.arturogutierrez.openticator.interactor;

import com.arturogutierrez.openticator.executor.PostExecutionThread;
import com.arturogutierrez.openticator.executor.ThreadExecutor;
import rx.Observable;
import rx.Observer;
import rx.Subscription;
import rx.schedulers.Schedulers;

public abstract class Interactor<T> {

  private final ThreadExecutor threadExecutor;
  private final PostExecutionThread postExecutionThread;

  private Subscription subscription;

  public Interactor(ThreadExecutor threadExecutor, PostExecutionThread postExecutionThread) {
    this.threadExecutor = threadExecutor;
    this.postExecutionThread = postExecutionThread;
  }

  public abstract Observable<T> createObservable();

  public void execute(Observer<T> interactorObserver) {
    subscription = createObservable().subscribeOn(Schedulers.from(threadExecutor))
        .observeOn(postExecutionThread.getScheduler())
        .subscribe(interactorObserver);
  }

  public void unsubscribe() {
    if (subscription != null && !subscription.isUnsubscribed()) {
      subscription.unsubscribe();
    }
  }
}

File: domain/src/main/java/com/arturogutierrez/openticator/interactor/Interactor.kt
package com.arturogutierrez.openticator.interactor

import com.arturogutierrez.openticator.executor.PostExecutionThread
import com.arturogutierrez.openticator.executor.ThreadExecutor
import rx.Observable
import rx.Observer
import rx.Subscription
import rx.schedulers.Schedulers

abstract class Interactor<T>(val threadExecutor: ThreadExecutor, val postExecutionThread: PostExecutionThread) {

    private var subscription: Subscription? = null

    abstract fun createObservable(): Observable<T>

    fun execute(interactorObserver: Observer<T>) {
        subscription = createObservable().subscribeOn(Schedulers.from(threadExecutor)).observeOn(postExecutionThread.scheduler).subscribe(interactorObserver)
    }

    fun unsubscribe() {
        if (subscription != null && !subscription!!.isUnsubscribed) {
            subscription!!.unsubscribe()
        }
    }
}
